networkx/algorithms/spanner.py

```diff
@@ -0,0 +1,341 @@
+
+    Returns
+    -------
+    int
+        The parameter k such that (at most) the desired stretch is
+        achieved.
+    """
+    return int(math.floor((stretch + 1) / 2))
+
+
+def _setup_residual_graph(G, weight):
+    """Setup residual graph as a copy of G with unique edges weights.
+
+    The node set of the residual graph corresponds to the set V' from
+    the Baswana-Sen paper and the edge set corresponds to the set E'
+    from the paper.
+
+    This function associates distinct weights to the edges of the
+    residual graph (even for unweighted input graphs), as required by
+    the algorithm.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        An undirected simple graph.
+
+    weight : object
+        The edge attribute to use as distance.
+
+    Returns
+    -------
+    NetworkX graph
+        The residual graph used for the Baswana-Sen algorithm.
+    """
+    residual_graph = G.copy()
+
+    # establish unique edge weights, even for unweighted graphs
+    for u, v in G.edges():
+        if not weight:
+            residual_graph[u][v]['weight'] = (id(u), id(v), 0)
+        else:
+            residual_graph[u][v]['weight'] = (G[u][v][weight], id(u), id(v))
+
+    return residual_graph
+
+
+def _lightest_edge_dicts(residual_graph, clustering, node):
+    """Find the lightest edge to each cluster.
+
+    Searches for the minimum-weight edge to each cluster adjacent to
+    the given node.
+
+    Parameters
+    ----------
+    residual_graph : NetworkX graph
+        The residual graph used by the Baswana-Sen algorithm.
+
+    clustering : dictionary
+        The current clustering of the nodes.
+
+    node : node
+        The node from which the search originates.
+
+    Returns
+    -------
+    lightest_edge_neighbor, lightest_edge_weight : dictionary, dictionary
+        lightest_edge_neighbor is a dictionary that maps a center C to
+        a node v in the corresponding cluster such that the edge from
+        the given node to v is the lightest edge from the given node to
+        any node in cluster. lightest_edge_weight maps a center C to the
+        weight of the aforementioned edge.
+
+    Notes
+    -----
+    If a cluster has no node that is adjacent to the given node in the
+    residual graph then the center of the cluster is not a key in the
+    returned dictionaries.
+    """
+    lightest_edge_neighbor = {}
+    lightest_edge_weight = {}
+    for neighbor in residual_graph.adj[node]:
+        neighbor_center = clustering[neighbor]
+        weight = residual_graph[node][neighbor]['weight']
+        if neighbor_center not in lightest_edge_weight or\
+                weight < lightest_edge_weight[neighbor_center]:
+            lightest_edge_neighbor[neighbor_center] = neighbor
+            lightest_edge_weight[neighbor_center] = weight
+    return lightest_edge_neighbor, lightest_edge_weight
+
+
+def _closest_sampled_center(lightest_edge_weight, neighboring_sampled_centers):
```

Comment:
Great suggestion, thanks! Not only does this simplify the code, it also allows me to address the issue you raised above in a different way: For unweighted graphs I can now set the weight in ```_setup_residual_graph``` to ```(id(u), id(v))``` instead of ```(id(u), id(v), 0)```.